jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletContainerInitializerHolder.java

```diff
@@ -0,0 +1,261 @@
+        }
+        finally
+        {
+            ctx.setExtendedListenerTypes(false);
+        }
+    }
+    
+    /**
+     * Re-inflate a stringified ServletContainerInitializerHolder.
+     * 
+     * @param loader the classloader to use to load the startup classes
+     * @param string the stringified representation of the ServletContainerInitializerHolder
+     * 
+     * @return a new ServletContainerInitializerHolder instance populated by the info in the string
+     */
+    public static ServletContainerInitializerHolder fromString(ClassLoader loader, String string)
+    {
+        Matcher m = __pattern.matcher(string);
+
+        if (!m.matches())
+            throw new IllegalArgumentException(string);
+
+        try
+        {
+            //load the ServletContainerInitializer and create an instance
+            String sciClassname = m.group(1);
+            ServletContainerInitializer sci = (ServletContainerInitializer)loader.loadClass(sciClassname).getDeclaredConstructor().newInstance();
+            ServletContainerInitializerHolder holder = new ServletContainerInitializerHolder(new Source(Source.Origin.ANNOTATION, sciClassname));
+            holder.setInstance(sci);
+            
+            //ensure all classes to be passed to onStartup are resolved
+            Set<Class<?>> classes = new HashSet<>();
+            String[] classnames = StringUtil.arrayFromString(m.group(2));
+            for (String name:classnames)
+                classes.add(loader.loadClass(name));
+            
+            classnames = StringUtil.arrayFromString(m.group(4));
+            for (String name:classnames)
+                classes.add(loader.loadClass(name));
+
+            classnames = StringUtil.arrayFromString(m.group(6));
+            for (String name:classnames)
+                classes.add(loader.loadClass(name));
+            
+            holder.addStartupClasses((Class<?>[])classes.toArray(new Class<?>[0]));
+            
+            return holder;
+        }
+        catch (Exception e)
+        {
+            throw new IllegalArgumentException(string, e);
+        }
+    }
+    
+    @Override
+    public String toString()
+    {
+        Set<String> interested = new HashSet<>(_startupClassNames);
+        _startupClasses.forEach((c) -> interested.add(c.getName()));
+        //for backward compatibility the old output format must be retained
+        return String.format("ContainerInitializer{%s,interested=%s,applicable=%s,annotated=%s}", getClassName(), interested, Collections.emptySet(), Collections.emptySet());
+    }
+
+    /**
+     * Experimental Wrapper mechanism for ServletContainerInitializer objects.
+     * <p>
+     * Beans in {@code ServletContextHandler} or {@code WebAppContext} that implement this interface
+     * will be called to optionally wrap any newly created ServletContainerInitializers
+     * (before their onStartup method is called)
+     * </p>
+     */
+    public interface WrapFunction
```

Comment:
This can be replaced by the standard `java.util.function.UnaryOperator<ServletContainerInitializer>`.

Reply:
I haven't invented this, I've just done exactly the same as is done for all of the other *Holder classes. If we change Wrapper to UnaryOperator then we change it for all *Holders, in a separate PR.